File: langs/nodejs/forgehx/src/forgeperm.js
// ForgePerm — 16-word ARX permutation (SPECIFICATION_X §4.3).

export const WORDS = 16;
export const ROUNDS = 8;

const MASK64 = 0xffffffffffffffffn;

export function rotl(x, n) {
  n &= 63;
  x &= MASK64;
  if (n === 0) return x;
  return ((x << BigInt(n)) | (x >> BigInt(64 - n))) & MASK64;
}

export function rotr(x, n) {
  n &= 63;
  x &= MASK64;
  if (n === 0) return x;
  return ((x >> BigInt(n)) | (x << BigInt(64 - n))) & MASK64;
}

function low32(x) {
  return x & 0xffffffffn;
}

export function roundConstant(round, index) {
  const x =
    (0x9e3779b97f4a7c15n ^
      (BigInt(round >>> 0) * 0xd1b54a32d192ed03n) ^
      (BigInt(index >>> 0) * 0xa24baed4963ee407n)) &
    MASK64;
  return rotl(x, (round + index * 3) & 63);
}

function quarterRound(s, ia, ib, ic, id) {
  let a = s[ia];
  let b = s[ib];
  let c = s[ic];
  let d = s[id];
  a = (a + b + 2n * low32(a) * low32(b)) & MASK64;
  d = rotr(d ^ a, 17);
  c = (c + d + 2n * low32(c) * low32(d)) & MASK64;
  b = rotr(b ^ c, 11);
  a = (a + b + 2n * low32(a) * low32(b)) & MASK64;
  d = rotr(d ^ a, 23);
  c = (c + d + 2n * low32(c) * low32(d)) & MASK64;
  b = rotr(b ^ c, 41);
  s[ia] = a;
  s[ib] = b;
  s[ic] = c;
  s[id] = d;
}

/** In-place ForgePerm over a 16-element BigInt[] state. */
export function permute(state) {
  if (state.length !== WORDS) {
    throw new Error("ForgePerm: state must be 16 words");
  }
  const temp = new Array(WORDS);
  for (let r = 0; r < ROUNDS; r++) {
    for (let i = 0; i < WORDS; i++) {
      state[i] = (state[i] ^ roundConstant(r, i)) & MASK64;
    }
    quarterRound(state, 0, 4, 8, 12);
    quarterRound(state, 1, 5, 9, 13);
    quarterRound(state, 2, 6, 10, 14);
    quarterRound(state, 3, 7, 11, 15);
    quarterRound(state, 0, 5, 10, 15);
    quarterRound(state, 1, 6, 11, 12);
    quarterRound(state, 2, 7, 8, 13);
    quarterRound(state, 3, 4, 9, 14);
    for (let i = 0; i < WORDS; i++) {
      temp[(i * 7 + 3) & 15] = state[i];
    }
    for (let i = 0; i < WORDS; i++) {
      state[i] = temp[i];
    }
  }
}

File: langs/rust/forgehx/src/forgeperm.rs
//! ForgePerm — 16-word ARX permutation (SPECIFICATION_X §4.3).

pub const WORDS: usize = 16;
pub const ROUNDS: usize = 8;

#[inline]
fn low32(x: u64) -> u64 {
    x & 0xFFFF_FFFF
}

#[inline]
pub fn rotl(x: u64, n: u32) -> u64 {
    x.rotate_left(n & 63)
}

#[inline]
fn rotr(x: u64, n: u32) -> u64 {
    x.rotate_right(n & 63)
}

#[inline]
fn round_constant(round: usize, index: usize) -> u64 {
    let r = round as u64;
    let i = index as u64;
    let x = 0x9E37_79B9_7F4A_7C15u64
        ^ r.wrapping_mul(0xD1B5_4A32_D192_ED03)
        ^ i.wrapping_mul(0xA24B_AED4_963E_E407);
    rotl(x, ((round + index * 3) & 63) as u32)
}

fn quarter_round(s: &mut [u64; WORDS], ia: usize, ib: usize, ic: usize, id: usize) {
    let mut a = s[ia];
    let mut b = s[ib];
    let mut c = s[ic];
    let mut d = s[id];

    a = a
        .wrapping_add(b)
        .wrapping_add(2u64.wrapping_mul(low32(a)).wrapping_mul(low32(b)));
    d = rotr(d ^ a, 17);
    c = c
        .wrapping_add(d)
        .wrapping_add(2u64.wrapping_mul(low32(c)).wrapping_mul(low32(d)));
    b = rotr(b ^ c, 11);
    a = a
        .wrapping_add(b)
        .wrapping_add(2u64.wrapping_mul(low32(a)).wrapping_mul(low32(b)));
    d = rotr(d ^ a, 23);
    c = c
        .wrapping_add(d)
        .wrapping_add(2u64.wrapping_mul(low32(c)).wrapping_mul(low32(d)));
    b = rotr(b ^ c, 41);

    s[ia] = a;
    s[ib] = b;
    s[ic] = c;
    s[id] = d;
}

/// In-place ForgePerm on a 16-word state (8 rounds).
pub fn permute(state: &mut [u64; WORDS]) {
    let mut temp = [0u64; WORDS];
    for r in 0..ROUNDS {
        for i in 0..WORDS {
            state[i] ^= round_constant(r, i);
        }
        quarter_round(state, 0, 4, 8, 12);
        quarter_round(state, 1, 5, 9, 13);
        quarter_round(state, 2, 6, 10, 14);
        quarter_round(state, 3, 7, 11, 15);
        quarter_round(state, 0, 5, 10, 15);
        quarter_round(state, 1, 6, 11, 12);
        quarter_round(state, 2, 7, 8, 13);
        quarter_round(state, 3, 4, 9, 14);
        for i in 0..WORDS {
            temp[(i * 7 + 3) & 15] = state[i];
        }
        *state = temp;
    }
}
